File: src/Engine/Grid2.cpp
#include "Grid2.h"

#include <cmath>

namespace BeitaGo {
	Grid2::Grid2() : Grid2(0, 0) {}

	Grid2::Grid2(Grid2::unit v) : Grid2(v, v) {}

	Grid2::Grid2(Grid2::unit x, Grid2::unit y) : _x(x), _y(y) {}

	Grid2 Grid2::operator-() const {
		return Grid2(-X(), -Y());
	}

	Grid2& Grid2::operator+=(const Grid2& o) {
		X() += o.X();
		Y() += o.Y();
		return *this;
	}

	Grid2& Grid2::operator-=(const Grid2& o) {
		return (*this += (-o));
	}

	Grid2 Grid2::operator+(const Grid2& o) const {
		return Grid2(*this) += o;
	}

	Grid2 Grid2::operator-(const Grid2& o) const {
		return Grid2(*this) -= o;
	}

	Grid2& Grid2::operator*=(Grid2::unit a) {
		X() *= a;
		Y() *= a;
		return *this;
	}

	Grid2& Grid2::operator/=(Grid2::unit a) {
		X() /= a;
		Y() /= a;
		return *this;
	}

	Grid2 Grid2::operator*(Grid2::unit a) const {
		return Grid2(*this) *= a;
	}

	Grid2 Grid2::operator/(Grid2::unit a) const {
		return Grid2(*this) /= a;
	}

	bool Grid2::operator==(const Grid2& o) const {
		return X() == o.X() && Y() == o.Y();
	}

	bool Grid2::operator!=(const Grid2& o) const {
		return !(*this == o);
	}

	Grid2::unit Grid2::X() const {
		return _x;
	}

	Grid2::unit Grid2::Y() const {
		return _y;
	}

	Grid2::unit& Grid2::X() {
		return _x;
	}

	Grid2::unit& Grid2::Y() {
		return _y;
	}

	Grid2::unit& Grid2::X(Grid2::unit x) {
		return _x = x;
	}

	Grid2::unit& Grid2::Y(Grid2::unit y) {
		return _y = y;
	}

	Grid2::unit Grid2::ManhattenDistance() const {
		return (X() < 0 ? -X() : X()) + (Y() < 0 ? -Y() : Y());
	}

	float Grid2::Magnitude() const {
		return std::sqrt(static_cast<float>(X() * X() + Y() * Y()));
	}
}

File: srcpy/engine/grid2.py
import math
from typing import TypeVar

T = TypeVar("T")


class Grid2:
	def __init__(self, x: int, y: int):
		self._x = x
		self._y = y

	def __str__(self) -> str:
		return "({}, {})".format(self._x,self. _y)

	def __neg__(self) -> TypeVar:
		return Grid2(-self._x, -self._y)
	
	def __iadd__(self, other: T) -> T:
		self._x += other.x()
		self._y += other.y()
		return self

	def __isub__(self, other: T) -> T:
		self._x -= other.x()
		self._y -= other.y()
		return self

	def __add__(self, other: T) -> T:
		return Grid2(self._x + other.x(), self._y + other.y())

	def __sub__(self, other: T) -> T:
		return Grid2(self._x - other.x(), self._y - other.y())

	def __imul__(self, other: T) -> T:
		self._x *= other
		self._y *= other
		return self

	def __ifloordiv__(self, other: T) -> T:
		self._x //= other
		self._y //= other
		return self

	def __mul__(self, other: T) -> T:
		return Grid2(self._x * other.x(), self._x * other.y())

	def __floordiv__(self, other: T) -> T:
		return Grid2(self._x // other.x(), self._y // other.y())

	def __eq__(self, other: T) -> bool:
		return self._x == other.x() and self._y == other.y()

	def __ne__(self, other: T) -> bool:
		return not self == other

	def x(self) -> int:
		return self._x

	def y(self) -> int:
		return self._y

	def manhatten_distance(self) -> int:
		return abs(self._x) + abs(self._y)

	def magnitude(self) -> float:
		return math.sqrt(self._x * self._x + self._y * self._y)

	def __iter__(self):
		return self

	def __hash__(self):
		return self._x * 92346902346 + self._y